### PR TITLE
Expose 'Modify' command in Rust/C/Java API 

### PR DESCRIPTION
**Pull request by mbudiu-vmw**
@lalithsuresh : is this sufficient for your needs? See the Java example for a use of the Modify API.
There are a bunch of other APIs we could consider adding, variants of these calls.

### CODE REVIEW

**Comment by lalithsuresh** [comment]
@mbudiu-vmw I'll give this a spin and let you know.

**Comment by ryzhyk** [comment]
Resolves #523

File: java/ddlogapi/DDlogRecord.java
Comment (lalithsuresh) [inline java/ddlogapi/DDlogRecord.java:406]: Better to call this getStructFieldUnchecked.

**Review by lalithsuresh — COMMENTED** [review]
I'll continue later after testing it.

File: java/ddlogapi/DDlogRecord.java
Comment (mbudiu-vmw) [reply java/ddlogapi/DDlogRecord.java:406]: It's checked, but not a checked exception.

File: java/ddlogapi.c
Comment (lalithsuresh) [inline java/ddlogapi.c:824]: might be cleaner to just use a goto to jump to the end of the function and call the frees you need.

**Review by ryzhyk — COMMENTED** [review]
Will read the rest and send more feedback later today.

**Comment by mbudiu-vmw** [comment]
I will wait for more feedback before sending a new version of the PR.

File: rust/template/ddlog.h
Comment (ryzhyk) [inline rust/template/ddlog.h:1564]: Not  a good choice of name, imo.  This argument describes a change to apply to the value.  `delta` coule be a better name, but it's already used with a different meaning in this API. Perhaps `patch`?

File: rust/template/differential_datalog/record.rs
Comment (ryzhyk) [inline rust/template/differential_datalog/record.rs:3]: Actually, Rust is correct, you should probably use an iterator, not a range loop.  I would do that rather than disable the warning.

**Review by ryzhyk — APPROVED** [review]

**Comment by lalithsuresh** [comment]
I'm running into some issues using this.

The ddlog program generated by the sql-ddlog compiler is:

```
import fp
import time
import sql
import sqlop

typedef Thosts = Thosts{id:Option<string>, capacity:Option<signed<64>>, up:Option<bool>}

input relation Rhosts[Thosts] primary key (row) (row.id)
output relation Rhostsv[Thosts]
output relation Rgood_hosts[Thosts]
Rhostsv[v0] :- Rhosts[v],var v0 = v.
Rgood_hosts[v0] :- Rhosts[v],unwrapBool(a_lt_NR(v.capacity, 64'sd10)),var v0 = v.
```

My update statement is: `update hosts set capacity = 11 where id = 'n1'`

My code to create the Modify command is like this:

```
final DDlogRecord key = matchExpressionFromWhere(where, pkFields, context);
final int numColumnsToUpdate = updateCall.getTargetColumnList().size();

final DDlogRecord[] updatedValues = new DDlogRecord[numColumnsToUpdate];
final String[] columnsToUpdate = new String[numColumnsToUpdate];
for (int i = 0; i < numColumnsToUpdate; i++) {
    final String columnName = ((SqlIdentifier) update.getTargetColumnList().get(i)).getSimple()
                                                                                    .toLowerCase();
    final Field<?> field = getFieldByName(allFields, columnName);
    final boolean isNullableField = field.getDataType().nullable();
    final DDlogRecord valueToUpdateTo = update.getSourceExpressionList().accept(PARSE_LITERALS);
    final DDlogRecord maybeWrapped = maybeOption(isNullableField, valueToUpdateTo);
    updatedValues[i] = maybeWrapped;
    columnsToUpdate[i] = columnName;
}
final DDlogRecord updateRecord = DDlogRecord.makeNamedStruct("", columnsToUpdate, updatedValues);
final int tableId = dDlogAPI.getTableId(ddlogRelationName(tableName));
final DDlogRecCommand command = new DDlogRecCommand(DDlogCommand.Kind.Modify, tableId, key, updateRecord);
dDlogAPI.applyUpdates(new DDlogRecCommand[]{command});
```

I'm noticing some odd behaviors.

1. When creating the DDlog Modify command, if I specify the field name as "CAPACITY" instead of "capacity", nothing crashes, but no row gets updated. So it silently fails.

2. When I correct the field name to "capacity", I get a different error depending on whether I pass "Some{11}" or "11" as the command.

When passing Some{11} (i.e., `maybeWrapped`):

```
Caused by: ddlogapi.DDlogException: ddlog_apply_updates(): error: Cannot use positional struct as mutator
	at ddlogapi.DDlogAPI.ddlog_apply_updates(Native Method)
	at ddlogapi.DDlogAPI.applyUpdates(DDlogAPI.java:497)
	at com.vmware.ddlog.DDlogJooqProvider$QueryVisitor.visitUpdate(DDlogJooqProvider.java:252)
	... 39 more
```

When passing 11 (i.e., `valueToUpdateTo`).

```
Caused by: ddlogapi.DDlogException: ddlog_apply_updates(): error: not a struct Int(BigInt { sign: Plus, data: BigUint { data: [11] } })
	at ddlogapi.DDlogAPI.ddlog_apply_updates(Native Method)
	at ddlogapi.DDlogAPI.applyUpdates(DDlogAPI.java:497)
	at com.vmware.ddlog.DDlogJooqProvider$QueryVisitor.visitUpdate(DDlogJooqProvider.java:252)
	... 39 more
```

The error message when I pass 11 makes sense, as the type does not match. So I suspect something is wrong about the behavior with passing Some{11}.

**Comment by lalithsuresh** [comment]
@mbudiu-vmw you can try my code above in the https://github.com/lalithsuresh/differential-datalog/tree/updates-with-ddlog-modify-command branch. The test to run isJooqProviderTest.testUpdates().

File: rust/template/ddlog.h
Comment (lalithsuresh) [reply rust/template/ddlog.h:1564]: Or modification.

**Comment by mbudiu-vmw** [comment]
@lalithsuresh hopefully this fixes the issue you have discovered.
It looks like my attempt to squash commits was a bit too successful - we only have the message from the previous commit, which I didn't make.

**Review by ryzhyk — COMMENTED** [review]
Looks great. Just need to untangle it from the old PR. Can I help with that?

**Comment by mbudiu-vmw** [comment]
@lalithsuresh : once this passes the CI tests we'll merge and hopefully you'll be good to go.

**Pull request merged**
